File: backend/users-ms/src/main/java/com/azat4dev/demobooking/users/users_commands/domain/core/values/LastName.java
package com.azat4dev.demobooking.users.users_commands.domain.core.values;

import com.azat4dev.demobooking.common.domain.DomainException;
import com.azat4dev.demobooking.common.utils.Assert;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.ToString;
import lombok.AccessLevel;


import java.io.Serializable;

@Getter
@EqualsAndHashCode
@ToString(of = "value")
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public final class LastName implements Serializable {

    public final static int MAX_LENGTH = 255;

    private final String value;

    private static void validate(String value) throws LastName.CantBeEmptyException, LastName.MaxLengthException {
        Assert.string(value, LastName.CantBeEmptyException::new).notNull().notBlank();
        Assert.string(value, LastName.MaxLengthException::new).maxLength(MAX_LENGTH);
    }

    public static LastName checkAndMakeFromString(String value) throws LastName.CantBeEmptyException, LastName.MaxLengthException {

        Assert.notNull(value, LastName.CantBeEmptyException::new);
        final var cleanedValue = value.trim();

        validate(cleanedValue);
        return new LastName(cleanedValue);
    }

    public static LastName dangerMakeFromStringWithoutCheck(String value) {
        assert value != null;
        return new LastName(value);
    }

    // Exceptions

    public static abstract class ValidationException extends DomainException {
        public ValidationException(String message) {
            super(message);
        }
    }

    public static final class CantBeEmptyException extends ValidationException {
        public CantBeEmptyException() {
            super("Last name cannot be empty");
        }

        @Override
        public String getCode() {
            return "Empty";
        }
    }

    public static final class MaxLengthException extends ValidationException {
        public MaxLengthException() {
            super("Last name is too long");
        }

        @Override
        public String getCode() {
            return "MaxLength";
        }

    }
}

File: web-client/src/domain/auth/values/LastName.ts
import Assert from "../../utils/Assert";
import ValidationException from "../interfaces/services/ValidationException";

class LastName {
    public static readonly MAX_LENGTH = 50;
    public static readonly MIN_LENGTH = 1;

    private constructor(public readonly value: string) {
    }

    private static validate =(value: string): void  =>{

        Assert.notBlank(value, () => new LastName.MinLengthException());
        Assert.minLength(
            value,
            LastName.MIN_LENGTH,
            () => new LastName.MinLengthException()
        );
        Assert.maxLength(
            value,
            LastName.MAX_LENGTH,
            () => new LastName.MaxLengthException()
        );
    }

    static dangerouslyCreate = (value: string): LastName => {
        return new LastName(value);
    }

    static checkAndCreate = (value: string): LastName => {
        LastName.validate(value);
        return new LastName(value);
    }

    static readonly ValidationException = ValidationException;

    static readonly MinLengthException = class extends LastName.ValidationException {
        public constructor() {
            super(
                `Last name must be at least ${LastName.MIN_LENGTH} characters long`
            );
        }

        getCode = (): string => {
            return "MinLength";
        };
    };

    static readonly MaxLengthException = class extends LastName.ValidationException {
        public constructor() {
            super(
                `Last name must be at most ${LastName.MAX_LENGTH} characters long`
            );
        }

        getCode = (): string => {
            return "MaxLength";
        };
    };
}

export default LastName;
